File: scanner/levels.py
"""Heuristic trade levels for a recommendation.

Mirror of web/src/lib/levels.ts — kept here so the desk's PM agent can write a
trade plan that references real entry/stop/target numbers, and so there's one
source of truth the frontend reads (rec["levels"]) instead of recomputing.

Output keys are camelCase to match the TS TradeLevels interface verbatim — the
dict is serialized straight into scan.json and consumed by the Svelte UI.

Method (see the TS file for the rationale):
  support (long)  = recent swing low, lifted to VWAP when VWAP is a nearer floor
  stop            = a volatility buffer (0.5 × ATR%) beyond that structure
  target          = max(recent swing high, 2R projection)
  rr              = reward ÷ risk
Shorts mirror it.  NOT investment advice.
"""
from __future__ import annotations


def _mean_abs_move(closes: list[float]) -> float:
    if len(closes) < 2:
        return 0.0
    total = 0.0
    n = 0
    for i in range(1, len(closes)):
        prev = closes[i - 1]
        if prev > 0:
            total += abs(closes[i] / prev - 1)
            n += 1
    return total / n if n else 0.0


def compute_levels(row: dict, side: str) -> dict | None:
    price = row.get("price")
    spark = row.get("spark")
    if price is None or price <= 0 or not spark or len(spark) < 5:
        return None

    # Nearer structure (last 5) anchors the stop so a far swing low on a runner
    # doesn't create an absurd 15%+ stop; wider window (last 10) reaches target.
    near_low = min(spark[-5:])
    near_high = max(spark[-5:])
    far_low = min(spark[-10:])
    far_high = max(spark[-10:])

    atr_pct = min(0.08, max(0.015, _mean_abs_move(spark)))  # 1.5%–8%
    buffer = price * atr_pct * 0.5
    max_risk = 0.08  # never risk more than ~8% to entry — keeps stops practical
    vwap = (row.get("intraday") or {}).get("vwap")

    if side == "long":
        support = near_low
        if vwap is not None and vwap < price and vwap > near_low:
            support = vwap
        if support >= price:
            support = price * (1 - atr_pct)
        stop = support - buffer
        min_stop = price * (1 - max_risk)
        if stop < min_stop:
            stop = min_stop
        risk = price - stop
        if risk <= 0:
            return None
        target = max(far_high, price + 2 * risk)
        rr = (target - price) / risk
        pivot, label = support, "support"
    else:
        resistance = near_high
        if vwap is not None and vwap > price and vwap < near_high:
            resistance = vwap
        if resistance <= price:
            resistance = price * (1 + atr_pct)
        stop = resistance + buffer
        max_stop = price * (1 + max_risk)
        if stop > max_stop:
            stop = max_stop
        risk = stop - price
        if risk <= 0:
            return None
        target = min(far_low, price - 2 * risk)
        rr = (price - target) / risk
        pivot, label = resistance, "resistance"

    return {
        "side": side,
        "entry": round(float(price), 2),
        "pivot": round(float(pivot), 2),
        "pivotLabel": label,
        "stop": round(float(stop), 2),
        "target": round(float(target), 2),
        "rr": round(float(rr), 2),
    }


def attach_levels(recommendations: dict, rows: list[dict]) -> None:
    """Set rec['levels'] on every pick, in place. Deterministic, no LLM."""
    by_ticker = {r["ticker"]: r for r in rows}
    for direction in ("longs", "shorts"):
        for rec in recommendations.get(direction, []):
            row = by_ticker.get(rec.get("ticker"))
            if not row:
                continue
            lv = compute_levels(row, rec.get("direction", "long"))
            if lv:
                rec["levels"] = lv

File: web/src/lib/levels.ts
// Heuristic trade levels for a recommendation, derived from the price action
// already in the scan row — no extra data fetch. These are NOT advice; they're
// a structural read of recent price: where support/resistance sit, a sensible
// stop beyond the structure, and a target projected from reward:risk.
//
// Inputs used (all already on ScanRow):
//   - price        current price (entry basis)
//   - spark        last ~20 daily closes → swing low/high + volatility proxy
//   - intraday     VWAP, HOD, LOD when in a live session (refines the level)
//
// Method:
//   support (long)  = nearest meaningful level below price: the recent swing
//                     low, lifted to VWAP if VWAP sits between it and price.
//   stop            = a volatility buffer beyond that structure (so noise
//                     doesn't stop you out). Buffer = 0.5 × ATR%, where ATR%
//                     is the mean absolute close-to-close move over the spark.
//   target          = the larger of the recent swing high and a 2R projection,
//                     so a breakout name isn't capped at a stale high.
//   rr              = reward ÷ risk.
// Shorts mirror this (resistance above, stop above, target below).

import type { ScanRow, TradeLevels, RiskSizing } from './types';

export type { TradeLevels, RiskSizing };

// Reference account equity for position sizing — mirrors scanner.main
// SIZING_REFERENCE_EQUITY (~£12k ISA). A yardstick for the suggested share
// count / concentration %, NOT a live brokerage balance.
export const REFERENCE_EQUITY = 12_000;

// A deterministic plain-language trade plan built from the levels — the
// fallback so there's ALWAYS written guidance on a pick, even before the
// desk's PM writes the richer version. Plain talk, real numbers, no LLM.
export function quickPlan(l: TradeLevels): string {
  const f = (n: number) => n.toFixed(2);
  const rr = l.rr != null ? `${l.rr.toFixed(1)}R` : '—';
  if (l.side === 'long') {
    const risk = (((l.entry - l.stop) / l.entry) * 100).toFixed(1);
    const rew = (((l.target - l.entry) / l.entry) * 100).toFixed(1);
    return `Enter near $${f(l.entry)}. Stop $${f(l.stop)} (−${risk}%); first target $${f(l.target)} (+${rew}%, ${rr}). Below support at $${f(l.pivot)} the setup's done.`;
  }
  const risk = (((l.stop - l.entry) / l.entry) * 100).toFixed(1);
  const rew = (((l.entry - l.target) / l.entry) * 100).toFixed(1);
  return `Short near $${f(l.entry)}. Stop $${f(l.stop)} (+${risk}%); first target $${f(l.target)} (−${rew}%, ${rr}). Above resistance at $${f(l.pivot)} the setup's done.`;
}

function meanAbsMove(closes: number[]): number {
  if (closes.length < 2) return 0;
  let sum = 0;
  let n = 0;
  for (let i = 1; i < closes.length; i++) {
    const prev = closes[i - 1];
    if (prev > 0) {
      sum += Math.abs(closes[i] / prev - 1);
      n++;
    }
  }
  return n ? sum / n : 0; // fractional (e.g. 0.03 = 3%)
}

export function computeLevels(row: ScanRow, side: 'long' | 'short'): TradeLevels | null {
  const price = row.price;
  const spark = row.spark;
  if (price == null || price <= 0 || !spark || spark.length < 5) return null;

  // Nearer structure (last 5 bars) anchors the stop, so a far swing low on a
  // name that's already run a long way doesn't create an absurd 15%+ stop.
  // Wider window (last 10) is only used to reach for the target.
  const nearLow = Math.min(...spark.slice(-5));
  const nearHigh = Math.max(...spark.slice(-5));
  const farLow = Math.min(...spark.slice(-10));
  const farHigh = Math.max(...spark.slice(-10));

  // Volatility buffer for the stop — clamp so it's neither hair-trigger nor huge.
  const atrPct = Math.min(0.08, Math.max(0.015, meanAbsMove(spark))); // 1.5%–8%
  const buffer = price * atrPct * 0.5;
  const MAX_RISK = 0.08; // never risk more than ~8% to entry — keeps stops practical

  const vwap = row.intraday?.vwap ?? null;

  if (side === 'long') {
    // Support: nearer swing low, lifted to VWAP if VWAP is a closer floor.
    let support = nearLow;
    if (vwap != null && vwap < price && vwap > nearLow) support = vwap;
    if (support >= price) support = price * (1 - atrPct);

    let stop = support - buffer;
    const minStop = price * (1 - MAX_RISK);
    if (stop < minStop) stop = minStop; // cap stop distance at MAX_RISK
    const risk = price - stop;
    if (risk <= 0) return null;
    const target = Math.max(farHigh, price + 2 * risk);
    const rr = (target - price) / risk;
    return { side, entry: price, pivot: support, pivotLabel: 'support', stop, target, rr };
  } else {
    // Resistance: nearer swing high, pulled to VWAP if VWAP is a closer ceiling.
    let resistance = nearHigh;
    if (vwap != null && vwap > price && vwap < nearHigh) resistance = vwap;
    if (resistance <= price) resistance = price * (1 + atrPct);

    let stop = resistance + buffer;
    const maxStop = price * (1 + MAX_RISK);
    if (stop > maxStop) stop = maxStop;
    const risk = stop - price;
    if (risk <= 0) return null;
    const target = Math.min(farLow, price - 2 * risk);
    const rr = (price - target) / risk;
    return { side, entry: price, pivot: resistance, pivotLabel: 'resistance', stop, target, rr };
  }
}

// Client-side mirror of scanner/risk.py size_position(). Fixed-fractional: risk
// `equity * riskPct` over the per-share stop distance |entry − stop|, then clamp
// so one position's notional never exceeds `equity * maxPositionPct` (capped=true
// when that bites). Returns null on degenerate inputs (non-finite, entry ≤ 0,
// entry == stop) so the card simply hides sizing. Direction-agnostic — only the
// absolute entry/stop distance matters, so it's identical for longs and shorts.
export function sizePosition(
  equity: number,
  entry: number,
  stop: number,
  riskPct = 0.0075,
  maxPositionPct = 0.25
): RiskSizing | null {
  if (![equity, entry, stop].every((n) => Number.isFinite(n))) return null;
  if (equity <= 0 || entry <= 0) return null;
  const dist = Math.abs(entry - stop);
  if (dist <= 0) return null;

  let shares = Math.floor((equity * riskPct) / dist);
  const maxNotional = equity * maxPositionPct;
  let capped = false;
  if (shares * entry > maxNotional) {
    shares = Math.floor(maxNotional / entry);
    capped = true;
  }
  if (shares < 0) shares = 0;

  const notional = shares * entry;
  return {
    shares,
    notional: Math.round(notional * 100) / 100,
    risk_amount: Math.round(shares * dist * 100) / 100,
    pct_of_equity: equity ? Math.round((notional / equity) * 10000) / 10000 : 0,
    capped
  };
}
